File: LeetCode/Easy/0112-path-sum/0112-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
    return hasPathSumWithLen(root, targetSum, 0)
}

func hasPathSumWithLen(root *TreeNode, targetSum int, size int) bool {
    if (root == nil) {
        return targetSum == 0 && size >= 1
    }
    if (root.Left != nil && root.Right == nil) {
        return hasPathSumWithLen(root.Left, targetSum - root.Val, size + 1)
    } else if (root.Left == nil && root.Right != nil) {
        return hasPathSumWithLen(root.Right, targetSum - root.Val, size + 1)
    } else {
        return hasPathSumWithLen(root.Left, targetSum - root.Val, size + 1) || hasPathSumWithLen(root.Right, targetSum - root.Val, size + 1)
    }
    
}

File: LeetCode/Easy/0112-path-sum/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }

        if (isLeafNode(root)) {
            return targetSum == root.val;
        }

        return hasPathSum(root.left, targetSum - root.val) ||
        hasPathSum(root.right, targetSum - root.val);
    }

    public boolean isLeafNode(TreeNode node) {
        return node != null && node.left == null && node.right == null;
    }
}